### PR TITLE
Allow to return ONNXProgram when calling `to_onnx(dynamo=True)`

### LINKED ISSUE #20810: Allow to return `ONNXProgram` when calling `to_onnx(dynamo=True)`
### Description & Motivation

Since torch 2.7.0, they started to unify the onnx export logic, thus when calling 
```python
torch.onnx.export(dynamo=True)
```
torch would start to return the `ONNXProgram` object. (https://github.com/pytorch/pytorch/pull/137296)

The `LightningModule` should also have the same behavior.

## Solution
```python
class LightningModule:
    @torch.no_grad()
    def to_onnx(
        self,
        file_path: Union[str, Path, BytesIO, None] = None,
        input_sample: Optional[Any] = None,
        **kwargs: Any,
    ) -> Union["ONNXProgram", None]:
        if kwargs.get("dynamo", False) and not _ONNXSCRIPT_AVAILABLE:
            raise ModuleNotFoundError(...)
        ...
        ret = torch.onnx.export(self, input_sample, file_path, **kwargs)
        self.train(mode)
        return ret
```

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GdoongMathew**
## What does this PR do?

Fixes #20810

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20811.org.readthedocs.build/en/20811/

### CODE REVIEW

**Comment by GdoongMathew** [comment]
Hi @Borda , a friendly ping for code review~

BTW, I'm wondering if there's any way to update the typing-extensions requirement during testing so that it could fit with onnxscript? Thanks.

File: src/lightning/pytorch/core/module.py [hunk 1]
```diff
@@ -1360,12 +1362,18 @@ def _verify_is_manual_optimization(self, fn_name: str) -> None:
             )
 
     @torch.no_grad()
-    def to_onnx(self, file_path: Union[str, Path, BytesIO], input_sample: Optional[Any] = None, **kwargs: Any) -> None:
+    def to_onnx(
+        self,
+        file_path: Union[str, Path, BytesIO, None] = None,
+        input_sample: Optional[Any] = None,
+        **kwargs: Any,
+    ) -> Union["ONNXProgram", None]:
```
Comment (deependujha) [inline src/lightning/pytorch/core/module.py:1370]: It's technically correct, but since we've only one type, what do you think about `Optional["ONNXProgram"]` for readability?

File: tests/tests_pytorch/models/test_onnx.py [hunk 2]
```diff
@@ -167,3 +167,13 @@ def to_numpy(tensor):
...
+
+
+@RunIf(onnx=True, min_torch="2.7.0", dynamo=True, onnxscript=True)
+def test_model_return_type():
+    model = BoringModel()
+    model.example_input_array = torch.randn((1, 32))
+    model.eval()
+
+    ret = model.to_onnx(dynamo=True)
+    assert isinstance(ret, torch.onnx.ONNXProgram)
```
[the remark this answers is not in the document]
Comment (GdoongMathew) [reply tests/tests_pytorch/models/test_onnx.py:186]: I think we could use the existed test `test_if_inference_output_is_valid` for this purpose? I added an parametrized `dynamo` to that unittest in the last commit as well.

Thank you ~.

**Comment by GdoongMathew** [comment]
> seems this to be the cause of issue:
> 
> https://github.com/Lightning-AI/pytorch-lightning/blob/dd2912a045f1a1a0c87528b41c9854fc7454d4b0/requirements/pytorch/base.txt#L10
> 
> I wonder what could be the reason behind this version pinning.
> 
> error:
> 
> ```
> The conflict is caused by:
>     pytorch-lightning 2.5.1rc2 depends on typing-extensions<4.14.0 and ==4.4.0
>     pytorch-lightning[extra,strategies,test] 2.5.1rc2 depends on typing-extensions<4.14.0 and ==4.4.0
>     lightning-utilities 0.10.0 depends on typing-extensions
>     torch 2.1.0 depends on typing-extensions
>     onnxscript 0.2.5 depends on typing_extensions>=4.10
>     onnxscript 0.2.4 depends on typing_extensions>=4.10
>     onnxscript 0.2.3 depends on typing_extensions>=4.10
>     onnxscript 0.2.2 depends on typing_extensions>=4.10
> ```

Yea, I think that's where the problem is, and there're two solutions for this. 

1. Update the requirement.
2. Figure out a way to only change the version to fit with `onnxscript` when testing.

Personally, I'd prefer the second one, but this probably would need some change to the ci config which I not familiar with.

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
not sure if the failing tests are related to this change or they are a bit flaky... 🤔

**Comment by GdoongMathew** [comment]
Hi @Borda
I think that's was coming from the updated version of `pydantic`. Normally during testing with the oldest pytorch, pydantic 1.x was installed, but in this PR, because of the onnxscript dependency, it needs typing-extensions 1.10 or later, which forces the ci to install pydantic 2.0, which causes the problem during importing deepspeed.
 That's only from looking at the log. If there's anything I could change in the ci setting, I'd love to test it out~

**Comment by GdoongMathew** [comment]
Or I'm wondering if it's possible to install onnxscript only when torch version 2.4.0 or above is installed during testing 🤔

**Comment by Borda** [comment]
> I'm wondering if it's possible to install onnxscript only when torch version 2.4.0 or above is installed during testing

Why this limitation? You can make some tests skip if PT is below a certain version...

File: requirements/fabric/base.txt [hunk 3]
```diff
@@ -4,5 +4,5 @@
 torch >=2.1.0, <2.8.0
 fsspec[http] >=2022.5.0, <2025.6.0
 packaging >=20.0, <=25.0
-typing-extensions >=4.4.0, <4.14.0
+typing-extensions >=4.10.0, <4.14.0
```
Comment (Borda) [inline requirements/fabric/base.txt]: So why do we need this bump?

Comment (GdoongMathew) [reply requirements/fabric/base.txt]: So this is governed by onnxscript [here](https://github.com/microsoft/onnxscript/blob/6edcfd5374ca8a3478ab3e97f022ee570aa993ad/pyproject.toml#L28), and I didn't think of a better solution to pass the tests with the oldest pytorch without changing the current ci.

Also mentioned in https://github.com/Lightning-AI/pytorch-lightning/pull/20811#issuecomment-2894858245

**Comment by GdoongMathew** [comment]
> > I'm wondering if it's possible to install onnxscript only when torch version 2.4.0 or above is installed during testing
> 
> Why this limitation? You can make some tests skip if PT is below a certain version...

Yes, currently only torch 2.7.0 or higher supports running the newly added tests. 
Also, onnxscript needs to be installed additionally as its not included in `torch`'s standard requirements - similar to `onnx`. Therefore, I suggest installing onnxscript only when the appropriate Torch version is installed.

**Comment by GdoongMathew** [comment]
Hi @Borda , mind if I pin `fastapi` to version `0.98.0` and `pydantic` to `1.10.22` to test if the unittest could pass?

File: .github/workflows/ci-tests-fabric.yml [hunk 4]
```diff
@@ -138,7 +138,7 @@ jobs:
       - name: Install package & dependencies
         timeout-minutes: 20
         run: |
-          pip install -e ".[${EXTRA_PREFIX}test,${EXTRA_PREFIX}strategies]" -U --prefer-binary \
+          pip install -e ".[${EXTRA_PREFIX}test,${EXTRA_PREFIX}strategies,${EXTRA_PREFIX}serve]" -U --prefer-binary \
```
Comment (GdoongMathew) [inline .github/workflows/ci-tests-fabric.yml]: Not sure how this should be changed so that it could install the right extra packages ...

**Review by Borda — APPROVED** [review]

File: requirements/pytorch/serve.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+fastapi >= 0.98.0
+pydantic >= 1.10.22
```
Comment (Borda) [inline requirements/pytorch/serve.txt]: I dint see these to be used anywhere

File: .github/workflows/ci-tests-fabric.yml (hunk 4 above)
Comment (Borda) [reply .github/workflows/ci-tests-fabric.yml]:
```suggestion
          pip install -e ".[${EXTRA_PREFIX}test,${EXTRA_PREFIX}strategies]" -U --prefer-binary \
```
